compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/ClassReference.kt

```diff
@@ -0,0 +1,162 @@
+  return fqName.classDescriptorOrNull(this)
+    ?.let { Descriptor(it, fqName) }
+    ?: getKtClassOrObjectOrNull(fqName)
+      ?.let { Psi(it, fqName) }
+}
+
+@ExperimentalAnvilApi
+public fun ClassDescriptor.toClassReference(): Descriptor {
+  return ClassReference.Descriptor(this, fqNameSafe)
+}
+
+@ExperimentalAnvilApi
+public fun KtClassOrObject.toClassReference(): Psi {
+  return ClassReference.Psi(this, requireFqName())
+}
+
+@ExperimentalAnvilApi
+public fun ModuleDescriptor.requireClassReference(fqName: FqName): ClassReference {
+  return classReferenceOrNull(fqName)
+    ?: throw AnvilCompilationException("Couldn't resolve ClassReference for $fqName")
+}
+
+/**
+ * This will return all super types as [ClassReference], whether they're parsed as [KtClassOrObject]
+ * or [ClassDescriptor]. This will include generated code, assuming it has already been generated.
+ * The returned sequence will be distinct by FqName, and Psi types are preferred over Descriptors.
+ *
+ * The first elements in the returned sequence represent the direct superclass to the receiver. The
+ * last elements represent the types which are furthest up-stream.
+ *
+ * @param includeSelf If true, the receiver class is the first element of the sequence
+ */
+@ExperimentalAnvilApi
+public fun ClassReference.allSuperTypeClassReferences(
+  module: ModuleDescriptor,
+  includeSelf: Boolean = false
+): Sequence<ClassReference> {
+
+  fun KtClassOrObject.superTypeEntryNames() = superTypeListEntries
+    .map { it.requireFqName(module) }
+
+  fun List<FqName>.typeRefs(): Sequence<ClassReference> {
+    return asSequence().map { module.requireClassReference(it) }
+  }
+
+  val initial = if (includeSelf) {
+    sequenceOf(this)
+  } else {
+    when (this) {
+      is Descriptor -> {
+        clazz.directSuperClassAndInterfaces()
+          .map { Descriptor(it, it.fqNameSafe) }
+          .asSequence()
+      }
+      is Psi -> {
+        clazz.superTypeEntryNames()
+          .typeRefs()
+      }
+    }
+  }
+
+  return generateSequence(initial) { superTypes ->
+    superTypes
+      .mapNotNull { classRef ->
+        when (classRef) {
+          is Psi -> {
+            classRef.clazz.superTypeEntryNames()
+              .takeIf { it.isNotEmpty() }
+              ?.typeRefs()
+          }
+          is Descriptor -> {
+            classRef.clazz.directSuperClassAndInterfaces()
+              .takeIf { it.isNotEmpty() }
+              ?.asSequence()
+              ?.map { module.requireClassReference(it.fqNameSafe) }
+          }
+        }
+      }
+      .flatten()
+      .takeIf { it.firstOrNull() != null }
+  }.flatten()
+    .distinctBy { it.fqName }
+}
+
+/**
+ * @param parameterName The name of the parameter to be found, not including any variance modifiers.
+ * @return The 0-based index of a declared generic type.
+ */
+@ExperimentalAnvilApi
+public fun ClassReference.indexOfTypeParameter(parameterName: String): Int {
+  return when (this) {
```

Comment:
I think instead of the when here, this should be a proper function on the type. This will avoid the `when` check.

Reply:
I can see how this looks a little weird in isolation.  

The idea is that `ClassReference` is the unifying facade, so that the `when` checks can be avoided elsewhere.  Those pending `BindingModuleGenerator` changes have more extension functions like this.  For instance, `ClassReference.scope(...)`, which just does a `when` and then delegates to the existing functions for the two different APIs.